_source/_posts/2022-09-15-kafka-microservices.md

```diff
@@ -0,0 +1,757 @@
+tags: [java, kafka, microservices, jhipster]
+tweets:
+- "Want to see how to use @apachekafka to communicate between microservices? This tutorial shows you how!"
+- "Did you know @jhipster supports #Kafka as a communications mechanism between microservices?"
+- "Microservice architectures ❤️ asynchronous messaging! Learn how to use @apachekafka to implement it. →"
+image: blog/kafka-microservices/kafka-microservices.png
+github: https://github.com/oktadev/okta-kafka-microservices-example
+type: conversion
+---
+
+One of the traditional approaches for communicating between [microservices](https://www.okta.com/blog/2021/02/microservices/) is through their REST APIs. However, as your system evolves and the number of microservices grows, communication becomes more complex, and the architecture might start resembling our old friend the spaghetti anti-pattern, with services depending on each other or tightly coupled, slowing down development teams. This model can exhibit low latency but only works if services are made highly available.
+
+To overcome this design disadvantage, new architectures aim to decouple senders from receivers, with asynchronous messaging. In a Kafka-centric architecture, low latency is preserved, with additional advantages like message balancing among available consumers and centralized management.
+
+When dealing with a brownfield platform (legacy), a recommended way to de-couple a monolith and ready it for a move to microservices is to implement asynchronous messaging.
+
+In this tutorial you will learn how to:
+
+- Create a microservices architecture with JHipster
+- Enable Kafka integration for communicating microservices
+- Set up Okta as the authentication provider
+
+{% include toc.md %}
+
+## What is Kafka?
+
+**Apache Kafka** is a distributed streaming platform. It was initially conceived as a message queue and open-sourced by LinkedIn in 2011. Its community evolved Kafka to provide key capabilities:
+
+- **Publish and Subscribe** to streams of records, like a message queue.
+- **Storage system** so messages can be consumed asynchronously. Kafka writes data to a scalable disk structure and replicates it for fault-tolerance. Producers can wait for write acknowledgments.
+- **Stream processing** with Kafka Streams API, enables complex aggregations or joins of input streams onto an output stream of processed data.
+
+Traditional messaging models are queue and publish-subscribe. In a queue, each record goes to one consumer. In publish-subscribe, the record is received by all consumers.
+
+The **Consumer Group** in Kafka is an abstraction that combines both models. Record processing can be load balanced among the members of a consumer group and Kafka allows to broadcast messages to multiple consumer groups. It is the same publish-subscribe semantic where the subscriber is a cluster of consumers instead of a single process.
```

Comment:
```suggestion
The **Consumer Group** in Kafka is an abstraction that combines both models. Record processing can be load balanced among the members of a consumer group and Kafka allows you to broadcast messages to multiple consumer groups. It is the same publish-subscribe semantic where the subscriber is a cluster of consumers instead of a single process.
```